### PR TITLE
[Flex_CheckPoint]: Support high-dim-reshard for flex_ckpt

### PR DESCRIPTION
**Pull request by Lmywl**
### PR Category

Auto Parallel

### PR Types

Bug fixes

### Description

Fix an incorrect minimal bounding slice computation in FlexCheckpoint (dcp) for logical shapes with **3 or more dimensions**, so that flattened parameters can be resharded correctly at any rank.

**Problem**

`minimal_nd_slice` (`python/paddle/distributed/flex_checkpoint/dcp/utils.py`) maps a flattened range `[flat_start, flat_end)` back to the minimal N-D slice that covers it. The old implementation only checked whether the **immediately preceding** axis had diverged:

```python
if start_idx[axis - 1] == end_idx[axis - 1]:
    s = min(start_idx[axis], end_idx[axis])
    e = max(start_idx[axis], end_idx[axis]) + 1
else:
    s, e = 0, shape[axis]
```

This happens to hold for 2-D, but not for 3-D and above: as soon as **any** higher axis diverges, the current axis must span its full range. When a higher axis diverges while the immediately preceding axis happens to be identical, the old logic wrongly narrows the current axis and the resulting slice no longer covers the whole flattened range.

Example with `shape = (2, 10, 12)` and range `[25, 155)`:

- `start_idx = (0, 2, 1)`, `end_idx = (1, 2, 10)`
- axis 0 → `(0, 2)`
- axis 1: `start_idx[0] != end_idx[0]`, so full range `(0, 10)`
- axis 2: `start_idx[1] == end_idx[1] == 2`, so wrongly narrowed to `(1, 11)`

Flat index 36, i.e. `(0, 3, 0)`, lies inside the range but its last-axis index 0 falls outside `(1, 11)`. Since the slice does not cover the real range, the offsets later derived by `flat_range_in_min_slice` are shifted as well, so loading reads wrong data or hits a shape mismatch.

**Change**

Track a cumulative `prefix_has_diverged` flag: once any higher axis has `start_idx[axis] != end_idx[axis]`, every following axis takes its full range; otherwise the axis is narrowed by min/max. Using min/max for axis 0 is equivalent to the previous special case (`flat_start < flat_end` guarantees `start_idx[0] <= end_idx[0]`), so all axes can now be handled by a single uniform branch.

**Tests**

Add `test/auto_parallel/hybrid_strategy/test_high_dim_reshard.py` and `high_dim_reshard_worker.py`, an end-to-end two-device test that validates `save_state_dict` / `load_state_dict` resharding with `GLOBAL_SHAPE = (2, 10, 12)`:

- Three sharding patterns: crossing axis boundaries (`cross_axis`), landing exactly on an axis boundary (`exact_axis_boundary`), and isolating a single element (`last_element`)
- Save and load use different shardings, so resharding is actually exercised
- Verified for all three comm methods: `broadcast`, `send_recv`, `grouped_send_recv`
- Element-wise comparison against expected values via `np.testing.assert_array_equal`

The test fails before the fix and passes after it.

### 是否引起精度变化

否

### CODE REVIEW

File: test/auto_parallel/hybrid_strategy/test_high_dim_reshard.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import collective.test_communication_api_base as test_base
+
+
+os.environ["FLAGS_enable_pir_api"] = "0"
+
+
+class TestHighDimReshard(test_base.CommunicationTestDistBase):
+    """End-to-end two-device FlexCheckpoint test for high-dimensional slices."""
+
+    def test_high_dim_reshard_two_devices(self):
```
Comment (risemeup1111) [inline test/auto_parallel/hybrid_strategy/test_high_dim_reshard.py:18]: P1 这个回归用例目前不会进入 CI。该目录的 Auto Parallel 用例由 `testslist.csv` 枚举执行，`CMakeLists.txt` 也由该表生成；当前两处都没有 `test_high_dim_reshard`，因此默认 CTest 和 `auto_unit_test` 都不会运行这里的断言。请将该用例加入 `testslist.csv`，并按 `CMakeLists.txt` 文件头的说明重新生成测试注册。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [test_high_dim_reshard.py:19](https://github.com/PaddlePaddle/Paddle/pull/79612#discussion_r3717646874) | P1 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: test/auto_parallel/hybrid_strategy/test_high_dim_reshard.py (hunk 1 above)
Comment (Lmywl) [reply test/auto_parallel/hybrid_strategy/test_high_dim_reshard.py:18]: done

Comment (risemeup1111) [reply test/auto_parallel/hybrid_strategy/test_high_dim_reshard.py:18]: 当前提交已将 `test_high_dim_reshard` 同步加入 `testslist.csv` 和生成的 `CMakeLists.txt`，测试入口已注册，此问题已修复。

**Review by risemeup1111 — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]

**Review by xingmingyyj — APPROVED** [review]
LGTM

**Review by From00 — APPROVED** [review]
LGTM

**Review by XieYunshen — APPROVED** [review]
